### PR TITLE
Add soxr resampler advanced configuration string like in piCorePlayer

### PR DESCRIPTION
**Pull request by crypto512**
…r soxr resampler.

For example from http://archimago.blogspot.com/2018/11/musings-raspberry-pi-3-b-touch.html:
```
resampler {
    plugin "soxr"
    quality "advanced"
    advanced_settings "v::3.05:28:99.7:100:45"
    threads "0"
}
```

### CODE REVIEW

**Comment by MaxKellermann** [comment]
Please add documentation

**Comment by crypto512** [comment]
@MaxKellermann done for the doc (i've limited to the minimum as the related documentation to understand the string is very verbose related to the current documentation).

**Comment by neheb** [comment]
This should be rebased to pass Travis.

**Comment by MaxKellermann** [comment]
I don't like linking to a random copy on some random Linux vendor website. This link will break as soon as Bionic manpages get removed because Bionic is outdated.

File: src/pcm/SoxrResampler.cxx [hunk 1]
```diff
@@ -22,12 +22,30 @@
...
+
 #include "Log.hxx"
 
 #include <soxr.h>
 
 #include <assert.h>
 #include <string.h>
+#include <math.h>
+
+struct soxr {
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Bad name. What does `soxr` mean?
Why not use `struct soxr_quality_spec`, which is essentially the same?

File: src/pcm/SoxrResampler.cxx [hunk 2]
```diff
@@ -22,12 +22,30 @@
...
+	unsigned long q_recipe;
+	unsigned long q_flags;
+	double q_precision;         /* Conversion precision (in bits).           20    */
+	double q_phase_response;    /* 0=minimum, ... 50=linear, ... 100=maximum 50    */
+	double q_passband_end;      /* 0dB pt. bandwidth to preserve; nyquist=1  0.913 */
+	double q_stopband_begin;    /* Aliasing/imaging control; > passband_end   1    */
+	double scale;
+	bool max_rate;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Not used

File: src/pcm/SoxrResampler.cxx [hunk 3]
```diff
@@ -22,12 +22,30 @@
...
+	bool exception;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Not used

File: src/pcm/SoxrResampler.cxx [hunk 4]
```diff
@@ -22,12 +22,30 @@
...
 #include "Log.hxx"
 
 #include <soxr.h>
 
 #include <assert.h>
 #include <string.h>
+#include <math.h>
+
+struct soxr {
+	unsigned long q_recipe;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: What is this `q_` prefix?

File: src/pcm/SoxrResampler.cxx [hunk 5]
```diff
@@ -22,12 +22,30 @@
...
+	double q_phase_response;    /* 0=minimum, ... 50=linear, ... 100=maximum 50    */
+	double q_passband_end;      /* 0dB pt. bandwidth to preserve; nyquist=1  0.913 */
+	double q_stopband_begin;    /* Aliasing/imaging control; > passband_end   1    */
+	double scale;
+	bool max_rate;
+	bool exception;
+};
+
+static struct soxr soxr_advanced_settings;
+char soxr_advanced_string[200];
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: What use is this? And why is it not static?

File: src/pcm/SoxrResampler.cxx [hunk 6]
```diff
@@ -50,6 +69,7 @@ static constexpr struct {
 	{ SOXR_MQ, "medium" },
 	{ SOXR_LQ, "low" },
 	{ SOXR_QQ, "quick" },
+	{ SOXR_32_BITQ, "advanced" },
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: What is `SOXR_32_BITQ` and why is it mapped to `advanced`?

File: src/pcm/SoxrResampler.cxx [hunk 7]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+static char *
+quality_advanced_nextparam(char *src, char c)
+{
+        static char *str = NULL;
+        char *ptr, *ret;
+
+        if (src)
+		str = src;
+        if (str && (ptr = strchr(str, c))) {
+                ret = str;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Bad indent

File: src/pcm/SoxrResampler.cxx [hunk 8]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
 	}
 }
 
+static char *
+quality_advanced_nextparam(char *src, char c)
+{
+        static char *str = NULL;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Static variable, oh noooo!

File: src/pcm/SoxrResampler.cxx [hunk 9]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+        char *ptr, *ret;
+
+        if (src)
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Huh, why this obscure `nullptr` check?

File: src/pcm/SoxrResampler.cxx [hunk 10]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
 	}
 }
 
+static char *
+quality_advanced_nextparam(char *src, char c)
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Undocumented obscure function

File: src/pcm/SoxrResampler.cxx [hunk 11]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+{
+        static char *str = NULL;
+        char *ptr, *ret;
+
+        if (src)
+		str = src;
+        if (str && (ptr = strchr(str, c))) {
+                ret = str;
+                *ptr = '\0';
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Why does it modify the input string?

File: src/pcm/SoxrResampler.cxx [hunk 12]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+                str = ptr + 1;
+        } else {
+                ret = str;
+                str = NULL;
+        }
+
+        return ret && ret[0] ? ret : NULL;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Very obscure. I really have no idea what this function is doing!

File: src/pcm/SoxrResampler.cxx [hunk 13]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+}
+
+static bool
+soxr_parse_quality_advanced(const char *copt, struct soxr *r)  noexcept
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Don't pass pointer arguments for the return value; just return the value!

File: src/pcm/SoxrResampler.cxx [hunk 14]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+{
+	char *opt, *recipe = NULL, *flags = NULL;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Don't use `NULL`, that's C and not C++!

File: src/pcm/SoxrResampler.cxx [hunk 15]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+	char *atten = NULL;
+	char *precision = NULL, *passband_end = NULL, *stopband_begin = NULL, *phase_response = NULL;
+
+	opt = xstrdup(copt);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Do not use `xstrdup()`, it's an error prone C function, leading to memory leaks.

File: src/pcm/SoxrResampler.cxx [hunk 16]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+		flags = quality_advanced_nextparam(NULL, ':');
+		atten = quality_advanced_nextparam(NULL, ':');
+		precision = quality_advanced_nextparam(NULL, ':');
+		passband_end = quality_advanced_nextparam(NULL, ':');
+		stopband_begin = quality_advanced_nextparam(NULL, ':');
+		phase_response = quality_advanced_nextparam(NULL, ':');
+	}
+
+	/* default to HQ (20 bit) if not user specified */
+	r->q_recipe = SOXR_HQ;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: So previously this was `SOXR_32_BITQ`, and now it's `SOXR_HQ`... huh?

File: src/pcm/SoxrResampler.cxx [hunk 17]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+	/* default to 1db of attenuation if not user specified */
+	r->scale = pow(10, -1.0 / 20);
+	/* override recipe derived values with user specified values */
+	r->q_precision = 0;
+	r->q_passband_end = 0;
+	r->q_stopband_begin = 0;
+	r->q_phase_response = -1;
+
+	if (recipe && recipe[0] != '\0') {
+		if (strchr(recipe, 'v')) r->q_recipe = SOXR_VHQ;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: So this checks only if a `v` is present, but allows lots of other garbage without complaining. Bad parser.

File: src/pcm/SoxrResampler.cxx [hunk 18]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+		if (strchr(recipe, 'h')) r->q_recipe = SOXR_HQ;
+		if (strchr(recipe, 'm')) r->q_recipe = SOXR_MQ;
+		if (strchr(recipe, 'l')) r->q_recipe = SOXR_LQ;
+		if (strchr(recipe, 'q')) r->q_recipe = SOXR_QQ;
+		if (strchr(recipe, 'L')) r->q_recipe |= SOXR_LINEAR_PHASE;
+		if (strchr(recipe, 'I')) r->q_recipe |= SOXR_INTERMEDIATE_PHASE;
+		if (strchr(recipe, 'M')) r->q_recipe |= SOXR_MINIMUM_PHASE;
+		if (strchr(recipe, 's')) r->q_recipe |= SOXR_STEEP_FILTER;
+		/* IGNORED: X = async resampling to max_rate */
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: I don't understand this comment.

File: src/pcm/SoxrResampler.cxx [hunk 19]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+		if (strchr(recipe, 'X')) r->max_rate = true;
+		/* IGNORED: E = exception, only resample if native rate is not */
+		if (strchr(recipe, 'E')) r->exception = true;
+	}
+
+	if (flags)
+		r->q_flags = strtoul(flags, 0, 16);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: No error checks.

File: src/pcm/SoxrResampler.cxx [hunk 20]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+
+	if (atten) {
+		double scale = pow(10, -atof(atten) / 20);
+		if (scale > 0 && scale <= 1.0)
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: Bad error check. It silently ignores invalid values.

File: src/pcm/SoxrResampler.cxx [hunk 21]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+			r->scale = scale;
+	}
+
+	if (precision)
+		r->q_precision = atof(precision);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: No error checks.

File: src/pcm/SoxrResampler.cxx [hunk 22]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+
+	if (passband_end)
+		r->q_passband_end = atof(passband_end) / 100;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: No error checks.

File: src/pcm/SoxrResampler.cxx [hunk 23]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+
+	if (stopband_begin)
+		r->q_stopband_begin = atof(stopband_begin) / 100;
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: No error checks.

File: src/pcm/SoxrResampler.cxx [hunk 24]
```diff
@@ -65,6 +85,104 @@ soxr_quality_name(unsigned long recipe) noexcept
...
+
+	if (phase_response)
+		r->q_phase_response = atof(phase_response);
```
Comment (MaxKellermann) [inline src/pcm/SoxrResampler.cxx]: No error checks.

**Comment by MaxKellermann** [comment]
I dug somewhat deeper and found out that you didn't write this code - you copy'n'pasted from https://github.com/ralph-irving/squeezelite/blame/master/resample.c
... which is GPLv3, but MPD is GPLv2.
Merging this would be a legal disaster.

It is a bit nefarious to copy other people's code and disguise it as your own work. And it is even worse to abuse this to force MPD to a stricter license. I'm happy I noticed, and didn't fall into that trap.
(This time I'm happy the code was so bad that I didn't even consider merging it.)

**Pull request closed without merging**